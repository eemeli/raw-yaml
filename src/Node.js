export const LOG = false

export class Range {
  constructor (start, end) {
    this.start = start
    this.end = end || start
  }
}

export default class Node {
  static Prop = {
    ANCHOR: 'ANCHOR',
    TAG: 'TAG'
  }

  constructor (src) {
    this.src = src
    this.nodeRange = null
    this.valueRange = null
    this.commentRange = null
    this.anchor = null
    this.tag = null
  }

  get comment () {
    if (!this.src || !this.commentRange) return null
    const { start, end } = this.commentRange
    return this.src.slice(start, end)
  }

  get rawValue () {
    if (!this.src || !this.valueRange) return null
    const { start, end } = this.valueRange
    return this.src.slice(start, end)
  }

  endIndent (offset) {
    let ch = this.src[offset]
    while (ch && ch === ' ') ch = this.src[offset += 1]
    return offset
  }

  endLine (offset) {
    let ch = this.src[offset]
    while (ch && ch !== '\n') ch = this.src[offset += 1]
    return offset
  }

  endWhiteSpace (offset) {
    let ch = this.src[offset]
    while (ch && (ch === '\t' || ch === ' ')) ch = this.src[offset += 1]
    return offset
  }

  parseComment (offset) {
    if (this.src[offset] === '#') {
      const start = offset + 1
      const end = this.endLine(start)
      this.commentRange = new Range(start, end)
      LOG && console.log('comment', this.commentRange)
      return end
    }
    LOG && console.log('comment skip')
    return offset
  }

  parseProps (offset) {
    let ch = this.src[offset]
    while (ch === '&' || ch === '!') {
      const end = this.endIdentifier(offset)
      const prop = ch === '&' ? 'anchor' : 'tag'
      this[prop] = this.src.slice(offset + 1, end)
      offset = this.endWhiteSpace(end)
      ch = this.src[offset]
      LOG && console.log('props', { prop, value: this[prop], offset, ch })
    }
    return offset
  }

}
