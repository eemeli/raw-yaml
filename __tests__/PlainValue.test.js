import Node from '../src/Node'
import { commonTests, testParse } from './common'

describe("parse single-line plain", () => {
  const type = Node.Type.PLAIN
  for (const name in commonTests) {
    const props = Object.assign({ str: 'plain value', type }, commonTests[name])
    test(name, () => testParse(props))
  }
  test('escaped', () => testParse({ pre: '? ', str: '-#:] ,', post: ': ', type, parent: { type: Node.Type.MAP_KEY } }))
  test('escaped in flow', () => testParse({ pre: '', str: '-#:', post: '] ,: ', inFlow: true, type }))
  test('empty', () => testParse({ pre: '- ', str: '', post: '\n- ', type }))
  test('empty in flow', () => testParse({ pre: '{ x: ', str: '', post: ' , ', inFlow: true, type }))
})

describe("parse multi-line plain", () => {
  const str = ':first#\n      ? :multi line\n  \n      \tplain value'
  const type = Node.Type.PLAIN
  for (const name in commonTests) {
    const props = Object.assign({ str, type }, commonTests[name])
    if (props.comment) {
      props.str = props.str.replace(/^.*/, '')
      props.expected = props.str.replace(/^\n/, '')
    }
    if (props.post && props.post[0] !== '\n') {
      props.pre = `  ${props.pre}`
      props.post = props.post.replace(/^\s?/, '\n')
    }
    test(name, () => testParse(props))
  }
})

describe('parse #comments', () => {
  const comment = 'comment # here!'
  test('bare', () => testParse({ pre: '', str: '', comment, post: '' }))
  test('newline before & after', () => testParse({ pre: '\n', str: '', comment, post: '\n' }))
  test('complex mapping key', () => testParse({ pre: '? ', str: '', comment, post: '\n: ', parent: { type: Node.Type.MAP_KEY } }))
  test('seq value', () => testParse({ pre: '- ', str: '', comment, post: '\n- ' }))
  test('indented block', () => testParse({ pre: '    - ', str: '', comment, post: '\n  x' }))
})
