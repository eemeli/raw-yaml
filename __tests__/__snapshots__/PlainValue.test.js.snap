// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse #comments bare 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 17,
    "start": 2,
  },
  "range": Range {
    "end": 17,
    "start": 0,
  },
  "raw": "",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 17,
    "start": 17,
  },
}
`;

exports[`parse #comments complex mapping key 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 19,
    "start": 4,
  },
  "range": Range {
    "end": 20,
    "start": 2,
  },
  "raw": "",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 20,
    "start": 19,
  },
}
`;

exports[`parse #comments indented block 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 23,
    "start": 8,
  },
  "range": Range {
    "end": 24,
    "start": 6,
  },
  "raw": "",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 24,
    "start": 23,
  },
}
`;

exports[`parse #comments newline before & after 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 18,
    "start": 3,
  },
  "range": Range {
    "end": 19,
    "start": 1,
  },
  "raw": "",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 19,
    "start": 18,
  },
}
`;

exports[`parse #comments seq value 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 19,
    "start": 4,
  },
  "range": Range {
    "end": 20,
    "start": 2,
  },
  "raw": "",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 20,
    "start": 19,
  },
}
`;

exports[`parse multi-line plain bare 1`] = `
PlainValue {
  "range": Range {
    "end": 49,
    "start": 0,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 49,
    "start": 0,
  },
}
`;

exports[`parse multi-line plain complex mapping key 1`] = `
PlainValue {
  "range": Range {
    "end": 54,
    "start": 4,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 54,
    "start": 4,
  },
}
`;

exports[`parse multi-line plain flow seq value 1`] = `
PlainValue {
  "range": Range {
    "end": 54,
    "start": 4,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 54,
    "start": 4,
  },
}
`;

exports[`parse multi-line plain indented block 1`] = `
PlainValue {
  "range": Range {
    "end": 56,
    "start": 6,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 56,
    "start": 6,
  },
}
`;

exports[`parse multi-line plain newline before & after 1`] = `
PlainValue {
  "range": Range {
    "end": 51,
    "start": 1,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 51,
    "start": 1,
  },
}
`;

exports[`parse multi-line plain seq value 1`] = `
PlainValue {
  "range": Range {
    "end": 52,
    "start": 2,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 52,
    "start": 2,
  },
}
`;

exports[`parse multi-line plain with comment 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 20,
    "start": 5,
  },
  "range": Range {
    "end": 63,
    "start": 3,
  },
  "raw": "? :multi line
  
      	plain value",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 63,
    "start": 20,
  },
}
`;

exports[`parse multi-line plain with props 1`] = `
PlainValue {
  "anchor": "anchor",
  "range": Range {
    "end": 66,
    "start": 2,
  },
  "raw": ":first#
      ? :multi line
  
      	plain value",
  "tag": "tag!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 66,
    "start": 16,
  },
}
`;

exports[`parse single-line plain bare 1`] = `
PlainValue {
  "range": Range {
    "end": 11,
    "start": 0,
  },
  "raw": "plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 11,
    "start": 0,
  },
}
`;

exports[`parse single-line plain complex mapping key 1`] = `
PlainValue {
  "range": Range {
    "end": 14,
    "start": 2,
  },
  "raw": "plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 14,
    "start": 2,
  },
}
`;

exports[`parse single-line plain empty 1`] = `
PlainValue {
  "range": Range {
    "end": 3,
    "start": 2,
  },
  "raw": "",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 3,
    "start": 2,
  },
}
`;

exports[`parse single-line plain empty in flow 1`] = `
PlainValue {
  "range": Range {
    "end": 6,
    "start": 5,
  },
  "raw": "",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 6,
    "start": 6,
  },
}
`;

exports[`parse single-line plain escaped 1`] = `
PlainValue {
  "range": Range {
    "end": 6,
    "start": 0,
  },
  "raw": "-#:] ,",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 6,
    "start": 0,
  },
}
`;

exports[`parse single-line plain escaped in flow 1`] = `
PlainValue {
  "range": Range {
    "end": 3,
    "start": 0,
  },
  "raw": "-#:",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 3,
    "start": 0,
  },
}
`;

exports[`parse single-line plain flow seq value 1`] = `
PlainValue {
  "range": Range {
    "end": 14,
    "start": 2,
  },
  "raw": "plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 14,
    "start": 2,
  },
}
`;

exports[`parse single-line plain indented block 1`] = `
PlainValue {
  "range": Range {
    "end": 18,
    "start": 6,
  },
  "raw": "plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 18,
    "start": 6,
  },
}
`;

exports[`parse single-line plain newline before & after 1`] = `
PlainValue {
  "range": Range {
    "end": 13,
    "start": 1,
  },
  "raw": "plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 13,
    "start": 1,
  },
}
`;

exports[`parse single-line plain seq value 1`] = `
PlainValue {
  "range": Range {
    "end": 14,
    "start": 2,
  },
  "raw": "plain value",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 14,
    "start": 2,
  },
}
`;

exports[`parse single-line plain with comment 1`] = `
PlainValue {
  "commentRange": Range {
    "end": 31,
    "start": 16,
  },
  "range": Range {
    "end": 31,
    "start": 3,
  },
  "raw": "plain value",
  "rawComment": "comment # here!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 14,
    "start": 3,
  },
}
`;

exports[`parse single-line plain with props 1`] = `
PlainValue {
  "anchor": "anchor",
  "range": Range {
    "end": 28,
    "start": 2,
  },
  "raw": "plain value",
  "tag": "tag!",
  "type": "PLAIN",
  "valueRange": Range {
    "end": 28,
    "start": 16,
  },
}
`;
